src/statsderl.erl

```diff
@@ -1,130 +1,309 @@
+  end.
 
--spec gauge_increment(key(), value(), sample_rate()) -> ok.
+%% @doc Maybe use env variables if the required arguments are not given.
+-spec maybe_use_env([{atom(), any()}]) -> [{atom(), any()}].
+maybe_use_env(Args) ->
+  maybe_use_env([{hostname, undefined} | ?DEFAULT_ARGS], Args).
 
-gauge_increment(Key, Value, SampleRate) when Value >= 0 ->
-    maybe_cast(gauge_increment, Key, Value, SampleRate).
+gauge(Key, Value, SampleRate) ->
+  gauge(?DEFAULT_REPORTER_NAME, Key, Value, SampleRate).
 
--spec increment(key(), value(), sample_rate()) -> ok.
+gauge(Reporter, Key, Value, SampleRate) ->
+  send_by_rate(Reporter, gauge, Key, Value, SampleRate).
 
-increment(Key, Value, SampleRate) when Value >= 0 ->
-    maybe_cast(counter, Key, Value, SampleRate).
+increment(Key, Value, SampleRate) ->
+  increment(?DEFAULT_REPORTER_NAME, Key, Value, SampleRate).
 
--spec timing(key(), value(), sample_rate()) -> ok.
+increment(Reporter, Key, Value, SampleRate) ->
+  send_by_rate(Reporter, increment, Key, Value, SampleRate).
 
 timing(Key, Value, SampleRate) ->
-    maybe_cast(timing, Key, Value, SampleRate).
+  timing(?DEFAULT_REPORTER_NAME, Key, Value, SampleRate).
 
--spec timing_fun(key(), fun(), sample_rate()) -> ok.
+timing(Reporter, Key, Value, SampleRate) ->
+  send_by_rate(Reporter, timing, Key, Value, SampleRate).
 
 timing_fun(Key, Fun, SampleRate) ->
-    Timestamp = statsderl_utils:timestamp(),
-    Result = Fun(),
-    timing_now(Key, Timestamp, SampleRate),
-    Result.
+  timing_fun(?DEFAULT_REPORTER_NAME, Key, Fun, SampleRate).
 
--spec timing_now(key(), erlang:timestamp(), sample_rate()) -> ok.
+timing_fun(Reporter, Key, Fun, SampleRate) ->
+  Timestamp = os:timestamp(),
+  Result = Fun(),
+  timing_now(Reporter, Key, Timestamp, SampleRate),
+  Result.
 
 timing_now(Key, Timestamp, SampleRate) ->
-    maybe_cast(timing_now, Key, Timestamp, SampleRate).
+  timing_now(?DEFAULT_REPORTER_NAME, Key, Timestamp, SampleRate).
+
+timing_now(Reporter, Key, Timestamp, SampleRate) ->
+  timing(Reporter, Key, now_diff_ms(Timestamp), SampleRate).
+
+
+%% @doc Report many metrics in one message.
+%% Same sample rate is used for all metrics in the batch
+%% @end
+-spec batch_report([{Method, Key, Value}], SampleRate) -> ok
+        when Method     :: increment | gauge | timing,
+             Key        :: iodata(),
+             Value      :: number(),
+             SampleRate :: number().
+batch_report(Metrics, SampleRate) ->
+  batch_report(?DEFAULT_REPORTER_NAME, Metrics, SampleRate).
+
+batch_report(Reporter, Metrics, SampleRate) ->
+  eval_by_rate(SampleRate,
+    fun() ->
+      MapFun = fun({Method, Key, Value}) ->
+                 iolist_to_binary(new_line(Method, Key, Value, SampleRate))
+               end,
+      Lines = lists:map(MapFun, Metrics),
+      gen_server:cast(Reporter, {send, Lines})
+    end).
+
+%%%_* gen_server callbacks =====================================================
+
+init(Args) ->
+  ArgF = fun(Name) ->
```

Comment:
It was mainly because the registration name has to be resolved before calling `gen_server:start_link`. (and in supervisor if it's configured in sys.config)
maybe_use_env is to have arg resolution (given > config > default) in one place for all.